In [2]:
from google.colab import files
uploaded=files.upload()

Saving diabetes.csv to diabetes.csv


In [4]:
import pandas as pd
df=pd.read_csv('diabetes.csv')
df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [5]:
#check null
df.isna().sum()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

In [6]:
df.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

In [7]:
X=df.drop('Outcome',axis='columns')

In [8]:
y=df['Outcome']

In [9]:
from sklearn.preprocessing import StandardScaler
scaler=StandardScaler()
X_scaled=scaler.fit_transform(X)

In [10]:
X_scaled

array([[ 0.63994726,  0.84832379,  0.14964075, ...,  0.20401277,
         0.46849198,  1.4259954 ],
       [-0.84488505, -1.12339636, -0.16054575, ..., -0.68442195,
        -0.36506078, -0.19067191],
       [ 1.23388019,  1.94372388, -0.26394125, ..., -1.10325546,
         0.60439732, -0.10558415],
       ...,
       [ 0.3429808 ,  0.00330087,  0.14964075, ..., -0.73518964,
        -0.68519336, -0.27575966],
       [-0.84488505,  0.1597866 , -0.47073225, ..., -0.24020459,
        -0.37110101,  1.17073215],
       [-0.84488505, -0.8730192 ,  0.04624525, ..., -0.20212881,
        -0.47378505, -0.87137393]])

In [11]:
from sklearn.model_selection import train_test_split

In [12]:
X_train,X_test,y_train,y_test=train_test_split(X_scaled,y,stratify=y,random_state=10)

In [13]:
X_train.shape

(576, 8)

In [14]:
X_test.shape

(192, 8)

In [17]:
y_train.value_counts()

0    375
1    201
Name: Outcome, dtype: int64

In [19]:
from sklearn.tree import DecisionTreeClassifier

In [20]:
from sklearn.model_selection import cross_val_score

In [22]:
scores=cross_val_score(DecisionTreeClassifier(),X,y,cv=5)

In [23]:
scores

array([0.69480519, 0.68181818, 0.70779221, 0.80392157, 0.75163399])

In [24]:
scores.mean()

0.727994227994228

In [26]:
from sklearn.ensemble import BaggingClassifier

In [28]:
bag=BaggingClassifier(
    base_estimator=DecisionTreeClassifier(),
    n_estimators=100,
    max_samples=0.8,
    oob_score=True,
    random_state=0

)

In [31]:
bag.fit(X_train,y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_base.py:166: FutureWarning: `base_estimator` was renamed to `estimator` in version 1.2 and will be removed in 1.4.
  warnings.warn(


BaggingClassifier(base_estimator=DecisionTreeClassifier(), max_samples=0.8,
                  n_estimators=100, oob_score=True, random_state=0)

In [33]:
bag.oob_score_

0.7534722222222222

### **Exercise**

In [35]:
from google.colab import files
uploaded=files.upload()
df=pd.read_csv('heart.csv')

Saving heart.csv to heart.csv


In [36]:
df.head()

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  MaxHR  \
0   40   M           ATA        140          289          0     Normal    172   
1   49   F           NAP        160          180          0     Normal    156   
2   37   M           ATA        130          283          0         ST     98   
3   48   F           ASY        138          214          0     Normal    108   
4   54   M           NAP        150          195          0     Normal    122   

  ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0              N      0.0       Up             0  
1              N      1.0     Flat             1  
2              N      0.0       Up             0  
3              Y      1.5     Flat             1  
4              N      0.0       Up             0

In [37]:
df[df.Cholesterol>(df.Cholesterol.mean()+3*df.Cholesterol.std())]

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
76    32   M           ASY        118          529          0     Normal   
149   54   M           ASY        130          603          1     Normal   
616   67   F           NAP        115          564          0        LVH   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
76     130              N      0.0     Flat             1  
149    125              Y      1.0     Flat             1  
616    160              N      1.6     Flat             0

In [38]:
df1 = df[df.Cholesterol<=(df.Cholesterol.mean()+3*df.Cholesterol.std())]
df1.shape

(915, 12)

In [39]:
df[df.MaxHR>(df.MaxHR.mean()+3*df.MaxHR.std())]

Empty DataFrame
Columns: [Age, Sex, ChestPainType, RestingBP, Cholesterol, FastingBS, RestingECG, MaxHR, ExerciseAngina, Oldpeak, ST_Slope, HeartDisease]
Index: []

In [40]:
df[df.FastingBS>(df.FastingBS.mean()+3*df.FastingBS.std())]

Empty DataFrame
Columns: [Age, Sex, ChestPainType, RestingBP, Cholesterol, FastingBS, RestingECG, MaxHR, ExerciseAngina, Oldpeak, ST_Slope, HeartDisease]
Index: []

In [41]:
df[df.Oldpeak>(df.Oldpeak.mean()+3*df.Oldpeak.std())]

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
166   50   M           ASY        140          231          0         ST   
702   59   M            TA        178          270          0        LVH   
771   55   M           ASY        140          217          0     Normal   
791   51   M           ASY        140          298          0     Normal   
850   62   F           ASY        160          164          0        LVH   
900   58   M           ASY        114          318          0         ST   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
166    140              Y      5.0     Flat             1  
702    145              N      4.2     Down             0  
771    111              Y      5.6     Down             1  
791    122              Y      4.2     Flat             1  
850    145              N      6.2     Down             1  
900    140              N      4.4     Down             1

In [42]:
df2 = df1[df1.Oldpeak<=(df1.Oldpeak.mean()+3*df1.Oldpeak.std())]
df2.shape

(909, 12)

In [43]:
df[df.RestingBP>(df.RestingBP.mean()+3*df.RestingBP.std())]

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
109   39   M           ATA        190          241          0     Normal   
241   54   M           ASY        200          198          0     Normal   
365   64   F           ASY        200            0          0     Normal   
399   61   M           NAP        200            0          1         ST   
592   61   M           ASY        190          287          1        LVH   
732   56   F           ASY        200          288          1        LVH   
759   54   M           ATA        192          283          0        LVH   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
109    106              N      0.0       Up             0  
241    142              Y      2.0     Flat             1  
365    140              Y      1.0     Flat             1  
399     70              N      0.0     Flat             1  
592    150              Y      2.0     Down             1  
732    133              Y      4.0     Down             1  
759    195              N      0.0       Up             1

In [44]:
df3 = df2[df2.RestingBP<=(df2.RestingBP.mean()+3*df2.RestingBP.std())]
df3.shape

(902, 12)

In [45]:
df4 = df3.copy()
df4.ExerciseAngina.replace(
    {
        'N': 0,
        'Y': 1
    },
    inplace=True)

df4.ST_Slope.replace(
    {
        'Down': 1,
        'Flat': 2,
        'Up': 3
    },
    inplace=True
)

df4.RestingECG.replace(
    {
        'Normal': 1,
        'ST': 2,
        'LVH': 3
    },
    inplace=True)

df4.head()

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS  RestingECG  \
0   40   M           ATA        140          289          0           1   
1   49   F           NAP        160          180          0           1   
2   37   M           ATA        130          283          0           2   
3   48   F           ASY        138          214          0           1   
4   54   M           NAP        150          195          0           1   

   MaxHR  ExerciseAngina  Oldpeak  ST_Slope  HeartDisease  
0    172               0      0.0         3             0  
1    156               0      1.0         2             1  
2     98               0      0.0         3             0  
3    108               1      1.5         2             1  
4    122               0      0.0         3             0

In [46]:
df5 = pd.get_dummies(df4, drop_first=True)
df5.head()

Age  RestingBP  Cholesterol  FastingBS  RestingECG  MaxHR  ExerciseAngina  \
0   40        140          289          0           1    172               0   
1   49        160          180          0           1    156               0   
2   37        130          283          0           2     98               0   
3   48        138          214          0           1    108               1   
4   54        150          195          0           1    122               0   

   Oldpeak  ST_Slope  HeartDisease  Sex_M  ChestPainType_ATA  \
0      0.0         3             0      1                  1   
1      1.0         2             1      0                  0   
2      0.0         3             0      1                  1   
3      1.5         2             1      0                  0   
4      0.0         3             0      1                  0   

   ChestPainType_NAP  ChestPainType_TA  
0                  0                 0  
1                  1                 0  
2                  0                 0  
3                  0                 0  
4                  1                 0

In [47]:
X = df5.drop("HeartDisease",axis='columns')
y = df5.HeartDisease

X.head()

Age  RestingBP  Cholesterol  FastingBS  RestingECG  MaxHR  ExerciseAngina  \
0   40        140          289          0           1    172               0   
1   49        160          180          0           1    156               0   
2   37        130          283          0           2     98               0   
3   48        138          214          0           1    108               1   
4   54        150          195          0           1    122               0   

   Oldpeak  ST_Slope  Sex_M  ChestPainType_ATA  ChestPainType_NAP  \
0      0.0         3      1                  1                  0   
1      1.0         2      0                  0                  1   
2      0.0         3      1                  1                  0   
3      1.5         2      0                  0                  0   
4      0.0         3      1                  0                  1   

   ChestPainType_TA  
0                 0  
1                 0  
2                 0  
3                 0  
4                 0

In [48]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)
X_scaled

array([[-1.42896269,  0.46089071,  0.85238015, ...,  2.06757196,
        -0.53547478, -0.22914788],
       [-0.47545956,  1.5925728 , -0.16132855, ..., -0.4836591 ,
         1.86750159, -0.22914788],
       [-1.74679706, -0.10495034,  0.79657967, ...,  2.06757196,
        -0.53547478, -0.22914788],
       ...,
       [ 0.37209878, -0.10495034, -0.61703246, ..., -0.4836591 ,
        -0.53547478, -0.22914788],
       [ 0.37209878, -0.10495034,  0.35947592, ...,  2.06757196,
        -0.53547478, -0.22914788],
       [-1.64085227,  0.3477225 , -0.20782894, ..., -0.4836591 ,
         1.86750159, -0.22914788]])

In [49]:

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=20)

In [50]:
from sklearn.svm import SVC
from sklearn.model_selection import cross_val_score

scores = cross_val_score(SVC(), X, y, cv=5)
scores.mean()

0.6906445672191528

In [51]:
from sklearn.ensemble import BaggingClassifier

bag_model = BaggingClassifier(base_estimator=SVC(), n_estimators=100, max_samples=0.8, random_state=0)
scores = cross_val_score(bag_model, X, y, cv=5)
scores.mean()

/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_base.py:166: FutureWarning: `base_estimator` was renamed to `estimator` in version 1.2 and will be removed in 1.4.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_base.py:166: FutureWarning: `base_estimator` was renamed to `estimator` in version 1.2 and will be removed in 1.4.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_base.py:166: FutureWarning: `base_estimator` was renamed to `estimator` in version 1.2 and will be removed in 1.4.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_base.py:166: FutureWarning: `base_estimator` was renamed to `estimator` in version 1.2 and will be removed in 1.4.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_base.py:166: FutureWarning: `base_estimator` was renamed to `estimator` in version 1.2 and will be removed in 1.4.
  warnings.warn(


0.6839656230816453

As you can see above, using bagging in case of SVM doesn't make much difference in terms of model accuracy. Bagging is effective when we have high variance and instable model such as decision tree. Let's explore how bagging changes the performance for a decision tree classifier.

In [52]:
from sklearn.tree import DecisionTreeClassifier

scores = cross_val_score(DecisionTreeClassifier(random_state=0), X, y, cv=5)
scores.mean()

0.7193984039287907

In [53]:
bag_model = BaggingClassifier(
    base_estimator=DecisionTreeClassifier(random_state=0),
    n_estimators=100,
    max_samples=0.9,
    oob_score=True,
    random_state=0
)

scores = cross_val_score(bag_model, X, y, cv=5)
scores.mean()

/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_base.py:166: FutureWarning: `base_estimator` was renamed to `estimator` in version 1.2 and will be removed in 1.4.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_base.py:166: FutureWarning: `base_estimator` was renamed to `estimator` in version 1.2 and will be removed in 1.4.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_base.py:166: FutureWarning: `base_estimator` was renamed to `estimator` in version 1.2 and will be removed in 1.4.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_base.py:166: FutureWarning: `base_estimator` was renamed to `estimator` in version 1.2 and will be removed in 1.4.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_base.py:166: FutureWarning: `base_estimator` was renamed to `estimator` in version 1.2 and will be removed in 1.4.
  warnings.warn(


0.8037016574585636

# **You can see that with bagging the score improved from 71.93% to 80.37%**

In [54]:
from sklearn.ensemble import RandomForestClassifier

scores = cross_val_score(RandomForestClassifier(), X, y, cv=5)
scores.mean()

0.8248004910988337